ember-file-upload/src/system/data-transfer-wrapper.ts

```diff
@@ -43,6 +47,81 @@ export default class DataTransferWrapper {
     return this.files.length ? this.files : this.items;
   }
 
+  async getFilesAndDirectories() {
+    const files: File[] = [];
+
+    const readEntry = async (entry: FileSystemEntry): Promise<File> => {
+      return new Promise((resolve, reject) => {
+        if (entry.isFile) {
+          (entry as FileSystemFileEntry).file((fileEntry: File) => {
+            resolve(fileEntry);
+          });
+        } else {
+          reject('Directory contains nested directories');
+        }
+      });
+    };
+
+    const getEntry = (
+      item: FutureProofDataTransferItem,
+    ): FileSystemDirectoryEntry => {
+      // In the future this method name might change, so already implementing it like this if needed
+      // https://developer.mozilla.org/en-US/docs/Web/API/DataTransferItem/webkitGetAsEntry
+      if (typeof item.getAsEntry === 'function') {
+        return item.getAsEntry();
+      }
+
+      return item.webkitGetAsEntry() as FileSystemDirectoryEntry;
+    };
+
+    const readAllFilesInDirectory = (item: DataTransferItem): Promise<File[]> =>
+      new Promise((resolve) => {
+        getEntry(item)
+          ?.createReader()
+          ?.readEntries(async (entries: FileSystemEntry[]) => {
+            const readFiles: File[] = await Promise.all(
+              entries.map(readEntry),
+            ).catch((err) => {
+              throw err;
+            });
+            resolve(readFiles.filter((file) => file !== undefined) as File[]);
```

Comment:
Safer to filter out missing elements with a [truthiness](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) check

```suggestion
            resolve(readFiles.filter(Boolean) as File[]);
```

Reply:
Didn't know you could use the `Boolean` constructor like this. Updated.